tools/cli/src/mentat_cli/command_parser.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright 2017 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use combine::{eof, many, many1, sep_by, skip_many, token, Parser};
+use combine::combinator::{choice, try};
+use combine::char::{alpha_num, space, string};
+
+pub static HELP_COMMAND: &'static str = &"help";
+pub static OPEN_COMMAND: &'static str = &"open";
+pub static CLOSE_COMMAND: &'static str = &"close";
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Command {
+    Transact(Vec<String>),
+    Query(Vec<String>),
+    Help(Vec<String>),
+    Open(String),
+    Close,
+    Err(String),
+}
+
+impl Command {
+    pub fn is_complete(&self) -> bool {
+        match self {
+            &Command::Query(_) |
+            &Command::Transact(_) => false,
+            _ => true
+        }
+    }
+}
+
+pub fn command(s: &str) -> Command {
+    let help_parser = string(HELP_COMMAND).and(skip_many(space())).and(sep_by::<Vec<_>, _, _>(many1::<Vec<_>, _>(alpha_num()), token(' '))).map(|x| {
+        let args: Vec<String> = x.1.iter().map(|v| v.iter().collect() ).collect();
+        Command::Help(args)
+    });
+
+    let open_parser = string(OPEN_COMMAND).and(space()).and(many1::<Vec<_>, _>(alpha_num()).and(eof())).map(|x| {
+        let arg: String = (x.1).0.iter().collect();
+        Command::Open(arg)
+    });
+
+    let close_parser = string(CLOSE_COMMAND).and(skip_many(space())).and(eof()).map( |_| Command::Close );
+
+    token('.')
+    .and(choice::<[&mut Parser<Input = _, Output = Command>; 3], _>
+        ([&mut try(help_parser),
+          &mut try(open_parser),
+          &mut try(close_parser),]))
```

Comment:
If you put `.skip(eof())` on the end here, you can remove it from the parsers above.